File: Social/Util/MentionHashtagText.swift
//
//  MentionHashtagText.swift
//  Social
//
//  Texto con "#etiquetas" Y "@menciones" tocables, compartido entre
//  captions de posts/reels y comentarios de posts/reels -- antes cada
//  pantalla que quería hashtags tocables tenía su propia copia de
//  `hashtagAttributedString`/`OpenURLAction` (HomeView.swift). A
//  diferencia de los sheets "Enviar a…" de esta sesión (duplicados a
//  propósito porque cada uno inserta contenido distinto en la base de
//  datos), aquí la lógica de renderizado es IDÉNTICA en las cuatro
//  superficies -- compartir es lo correcto, no duplicar. Equivalente de
//  MentionHashtagText.kt.
//
//  Nombre de usuario único real (@handle, 0073_profile_username.sql) +
//  notificación real de mención (0074_mentions.sql), comparado con
//  Instagram/Twitter/TikTok.
//

import SwiftUI

/// Construida FUERA de `body` a propósito, mismo criterio que
/// `hashtagAttributedString` (HomeView.swift): `body` está sujeto a
/// `@ViewBuilder` (requisito del protocolo `View`), que no admite un `for`
/// con efectos secundarios sin producir una `View` en cada iteración -- una
/// función normal fuera de cualquier result builder no tiene esa
/// restricción.
private func mentionHashtagAttributedString(_ text: String, baseColor: Color?, linkColor: Color) -> AttributedString {
    var result = AttributedString("")
    let words = text.split(separator: " ", omittingEmptySubsequences: false)
    for (index, word) in words.enumerated() {
        var piece = AttributedString(word)
        var isLink = false
        if word.hasPrefix("#"), word.count > 1 {
            let tag = word.dropFirst().trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
            if !tag.isEmpty, let url = URL(string: "socialhashtag://\(tag)") {
                piece.link = url
                piece.foregroundColor = linkColor
                piece.underlineStyle = .single
                isLink = true
            }
        } else if word.hasPrefix("@"), word.count > 1 {
            let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "_"))
            let handle = word.dropFirst().trimmingCharacters(in: allowed.inverted)
            if !handle.isEmpty, let url = URL(string: "socialmention://\(handle)") {
                piece.link = url
                piece.foregroundColor = linkColor
                piece.underlineStyle = .single
                isLink = true
            }
        }
        if !isLink, let baseColor {
            piece.foregroundColor = baseColor
        }
        result += piece
        if index != words.count - 1 { result += AttributedString(" ") }
    }
    return result
}

struct MentionHashtagText: View {
    let text: String
    var font: Font = .subheadline
    // Reels (ReelsView.swift) pinta este texto en blanco fijo sobre el
    // vídeo, no el color normal del resto de la app -- de ahí este
    // override opcional, en vez de forzar un único esquema de color para
    // las cuatro superficies. `nil` deja el texto normal heredar el color
    // ambiente, igual que antes de esta ronda.
    var color: Color? = nil
    var linkColor: Color = .accentColor
    var onOpenHashtag: (String) -> Void = { _ in }
    var onOpenMention: (String) -> Void = { _ in }

    var body: some View {
        Text(mentionHashtagAttributedString(text, baseColor: color, linkColor: linkColor))
            .font(font)
            .environment(\.openURL, OpenURLAction { url in
                if url.scheme == "socialhashtag", let tag = url.host {
                    onOpenHashtag(tag)
                    return .handled
                }
                if url.scheme == "socialmention", let handle = url.host {
                    onOpenMention(handle)
                    return .handled
                }
                return .systemAction
            })
    }
}

File: Android/app/src/main/java/com/social/app/util/MentionHashtagText.kt
package com.social.app.util

import androidx.compose.foundation.text.ClickableText
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.LocalTextStyle
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.text.style.TextDecoration

/**
 * Texto con "#etiquetas" Y "@menciones" tocables, compartido entre
 * captions de posts/reels y comentarios de posts/reels -- antes cada
 * pantalla que quería hashtags tocables tenía su propia copia de
 * `buildAnnotatedStringWithHashtags`/`CaptionText` (HomeScreen.kt). A
 * diferencia de los sheets "Enviar a…" de esta sesión (duplicados a
 * propósito porque cada uno inserta contenido distinto en la base de
 * datos), aquí la lógica de renderizado es IDÉNTICA en las cuatro
 * superficies -- compartir es lo correcto, no duplicar.
 *
 * Nombre de usuario único real (@handle, 0073_profile_username.sql) +
 * notificación real de mención (0074_mentions.sql), comparado con
 * Instagram/Twitter/TikTok.
 */
private fun buildMentionHashtagAnnotatedString(text: String, linkColor: Color, baseColor: Color): AnnotatedString {
    return buildAnnotatedString {
        val words = text.split(" ")
        words.forEachIndexed { index, word ->
            when {
                word.startsWith("#") && word.length > 1 -> {
                    val tag = word.drop(1).trimEnd { !it.isLetterOrDigit() }
                    pushStringAnnotation(tag = "hashtag", annotation = tag)
                    withStyle(SpanStyle(color = linkColor, textDecoration = TextDecoration.Underline)) { append(word) }
                    pop()
                }
                word.startsWith("@") && word.length > 1 -> {
                    val handle = word.drop(1).trimEnd { !(it.isLetterOrDigit() || it == '_') }
                    if (handle.isNotEmpty()) {
                        pushStringAnnotation(tag = "mention", annotation = handle)
                        withStyle(SpanStyle(color = linkColor, textDecoration = TextDecoration.Underline)) { append(word) }
                        pop()
                    } else {
                        withStyle(SpanStyle(color = baseColor)) { append(word) }
                    }
                }
                else -> withStyle(SpanStyle(color = baseColor)) { append(word) }
            }
            if (index != words.lastIndex) append(" ")
        }
    }
}

@Composable
fun MentionHashtagText(
    text: String,
    modifier: Modifier = Modifier,
    style: TextStyle = LocalTextStyle.current,
    // Reels (ReelsScreen.kt) pinta este texto en blanco fijo sobre el
    // vídeo, no el LocalContentColor/primary normales del resto de la
    // app -- de ahí estos dos overrides opcionales, en vez de forzar un
    // único esquema de color para las cuatro superficies.
    baseColor: Color = LocalContentColor.current,
    linkColor: Color = MaterialTheme.colorScheme.primary,
    onOpenHashtag: (String) -> Unit = {},
    onOpenMention: (String) -> Unit = {}
) {
    val annotated = remember(text, baseColor, linkColor) { buildMentionHashtagAnnotatedString(text, linkColor, baseColor) }
    ClickableText(text = annotated, modifier = modifier, style = style) { offset ->
        val hashtag = annotated.getStringAnnotations(tag = "hashtag", start = offset, end = offset).firstOrNull()
        if (hashtag != null) {
            onOpenHashtag(hashtag.item)
            return@ClickableText
        }
        annotated.getStringAnnotations(tag = "mention", start = offset, end = offset).firstOrNull()
            ?.let { onOpenMention(it.item) }
    }
}
